app/src/main/java/com/github/sdpcoachme/firebase/database/FireDatabase.kt

```diff
@@ -52,14 +51,31 @@ class FireDatabase(databaseReference: DatabaseReference) : Database {
         currEmail = email
     }
 
+    /**
+     * Gets all children of a given database reference
+     * @param databaseRef the database reference whose children to get
+     * @return a completable future that completes when the children are retrieved. The future
+     * contains a map of the children, with the key being the key of the child and the value being
+     * the child itself
+     */
+    private fun getAllChildren(databaseRef: DatabaseReference): CompletableFuture<Map<String, DataSnapshot>> {
+        return getRef(databaseRef).thenApply {
+            it.children.associateBy { child -> child.key!! /* can't be null */ }
+        }
+    }
+
     /**
      * Sets a key-value pair with a given key in a given database reference
      * @param databaseChild the database reference in which to set the key-value pair
      * @param key the key of the value
      * @param value the value to set
      * @return a completable future that completes when the child is set
+     * @throws IllegalArgumentException if the key is empty (no child can have an empty key)
      */
     private fun setChild(databaseChild: DatabaseReference, key: String, value: Any): CompletableFuture<Void> {
+        // Careful here, giving an empty string as a key will return a reference to the parent!!
+        if (key.isEmpty())
+            throw IllegalArgumentException("Key cannot be empty")
```

Comment:
This is a very important modification, as giving an empty key was triggering a much wrong behaviour before, instead of throwing an exception.
Nice.